### PR TITLE
feat(fenced-code-language): Add option to refuse extra data

### PR DESCRIPTION
**Pull request by regseb**
Add option to refuse leading and trailing whitespace; and extra data in fence code blocks information.

PR related:  #628 and #610

### CODE REVIEW

File: demo/markdownlint-browser.js [hunk 1]
```diff
@@ -4176,6 +4178,9 @@ module.exports = {
             else if ((allowed.length > 0) && !allowed.includes(lang)) {
                 addError(onError, token.lineNumber, `"${lang}" is not allowed`);
             }
+            if (!extra && token.info !== lang) {
```
Comment (DavidAnson) [inline demo/markdownlint-browser.js:4181]: Parenthesis, please.

File: demo/markdownlint-browser.js [hunk 2]
```diff
@@ -4168,6 +4168,8 @@ module.exports = {
     "function": function MD040(params, onError) {
         let allowed = params.config.allowed_languages;
         allowed = Array.isArray(allowed) ? allowed : [];
+        let extra = params.config.extra;
+        extra = (extra === undefined) ? true : !!extra;
```
Comment (DavidAnson) [inline demo/markdownlint-browser.js:4172]: I prefer to default unset parameters to false. Could this be reframed as "language_only"? (Not sure I love that name, but you get the idea.)

File: doc-build/md040.md [hunk 3]
```diff
@@ -29,6 +29,12 @@ You can configure the `allowed_languages` parameter to specify a list of
 languages code blocks could use. Languages are case sensitive. The default value
 is `[]` which means any language specifier is valid.
 
+You have the option to refuse extra data in fence code blocks information. To do
```
Comment (DavidAnson) [inline doc-build/md040.md:32]: I may have future comments on the phrasing here, but will wait until we agree on the parameter name.

File: test/md040-extra.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# md040-extra.md
+
+Fence code block information with leading whitespaces:
+
+```  html
+<h1>markdownlint</h1>  {MD040:5}
```
Comment (DavidAnson) [inline test/md040-extra.md:6]: Nit: Extra space

File: test/md040-extra.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# md040-extra.md
```
Comment (DavidAnson) [inline test/md040-extra.md:1]: Love the new test cases!

File: test/md040-extra.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+Fence code block information with leading whitespaces:
+
+```  html
+<h1>markdownlint</h1>  {MD040:5}
+```
+
+Fence code block information with trailing whitespace:
+
+```css 
+body {}  {MD040:11} {MD009:11}
```
Comment (DavidAnson) [inline test/md040-extra.md:12]: It's not unreasonable, but should we really warn for both rules in this scenario? Per the specification, trailing white space will already be removed and so is harmless and arguably not "extra" data. I'm thinking maybe only the trailing white space rule should report for this example. What do you think?

**Review by DavidAnson — CHANGES_REQUESTED** [review]
I think this is a good idea, and I like the behavior you have implemented! I left a few comments about how this PR might be tweaked. Thanks!

**Comment by regseb** [comment]
My proposals for the name of the parameter (from best to worst): `strict`, `language_only`, `no_extra`, `no_extra_data`, `one_word`, `single`.

---

With the `language_only` option, I expect the first line of the fenced code block to contain only the language (without spaces).

The two rules `no-trailing-spaces` and `fenced-code-language` check the same thing, but they don't have the same scope. Maybe a user only wants to check the language strictly. If we don't check the trailing spaces in the _info string_: the user won't get the information about the extra spaces.

The best thing to do would be to add a parameter to the `no-trailing-spaces` rule to not check for trailing spaces for fences. Personally, I wouldn't activate this parameter because I don't mind having two errors for the same case. I think this option would be superfluous.

**Comment by DavidAnson** [comment]
I don't like `strict` for this, so let's go with `language_only` - and it looks like you've already made that change. I'm fine with the overlap in trailing spaces we have now. I'll make a second review pass shortly. Thanks!

File: lib/md040.js [hunk 7]
```diff
@@ -11,6 +11,8 @@ module.exports = {
   "function": function MD040(params, onError) {
     let allowed = params.config.allowed_languages;
     allowed = Array.isArray(allowed) ? allowed : [];
+    const only = !!params.config.language_only;
```
Comment (DavidAnson) [inline lib/md040.js:14]: Please call it `languageOnly` for consistency.

File: lib/md040.js [hunk 8]
```diff
@@ -22,6 +24,14 @@ module.exports = {
           `"${lang}" is not allowed`
         );
       }
+
+      if (only && (token.info !== lang)) {
+        addError(
+          onError,
+          token.lineNumber,
+          `"${token.info}" doesn't only contain the language`
```
Comment (DavidAnson) [inline lib/md040.js:32]: Info string contains more than language: "${token.info}"

File: schema/build-config-schema.js [hunk 9]
```diff
@@ -356,6 +356,11 @@ for (const rule of rules) {
             "type": "string"
           },
           "default": []
+        },
+        "language_only": {
+          "description": "Accept language only",
```
Comment (DavidAnson) [inline schema/build-config-schema.js:361]: "Accept" -> "Require"

File: test/md040-language_only.md [hunk 10]
```diff
@@ -0,0 +1,25 @@
+# md040-language_only.md
+
+Fence code block information with leading whitespaces:
```
Comment (DavidAnson) [inline test/md040-language_only.md:3]: "whitespace" (no 's')

File: doc-build/md040.md (hunk 3 above)
Comment (DavidAnson) [reply doc-build/md040.md:32]: How about:

You can prevent extra data from being present in the info string of fenced code blocks. To do so, set the `language_only` parameter to `true`.
 
 Info strings with leading/trailing whitespace (ex: `  js  `) or other content (ex: `ruby startline=3`) will trigger this rule.

**Review by DavidAnson — CHANGES_REQUESTED** [review]
Looking good, just a few more comments...

File: test/md040-language_only.md [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+```  html
+<h1>markdownlint</h1> {MD040:5}
+```
+
+Fence code block information with trailing whitespace:
+
+```css 
+body {}  {MD040:11} {MD009:11}
```
Comment (DavidAnson) [inline test/md040-language_only.md:12]: Extra space

**Comment by regseb** [comment]
I fixed your feedback. For the text in English, I trust you. _(Google Translate est mon ami)_

**Review by DavidAnson — APPROVED** [review]
Thank you!

**Pull request merged**